### PR TITLE
Implement calldata decoding

### PR DESCRIPTION
**Pull request by haltman-at**
This PR implements decoding for variables declared as `calldata`.  Note that what's in this PR is actually a little more general than what can presently be tested, as I wrote a decoder for everything we can expect to eventually go in calldata, while as of Solidity 0.5.4, types with elements of dynamic type are still not allowed in calldata.  Still, given how much can already go in calldata, it was easier to just implement a full calldata decoder rather than try to restrict it to what can currently be tested.

Doing this can be split into three essential parts.  Firstly, reading and decoding calldata.  I don't have much to say about this, as it's basically a straightforward implementation of the ABI specification.  (For reading I didn't even make a separate function, but just reused `readBytes` from `read/memory.ts`; there is no `read/calldata.ts` as a result.)  The one thing of note is that some of the calldata decoding functions take an additional `base` parameter which is the base that relative pointers should be considered relative to.  Also, when decoding pointers to calldata from the stack for dynamic-length objects, we actually throw away the second word that contains the length, and instead use a base of `-DecodeUtils.EVM.WORD_SIZE` (this is labeled as a hack :P ) to read the length from calldata instead, since this allows us to treat things more consistently.

Secondly, computing sizes and allocating structs in calldata -- like in storage, we preallocate calldata structs, and we often have need to compute sizes of types.  So, there's now a calldata allocator, similar to the storage allocator but much simpler; go read what I wrote when I rewrote the storage allocator if you want to know more about that.  However, in calldata, in addition to the raw size of a type, we also care about whether or not that type is *dynamic* (in the sense that the ABI spec uses that term).  So, `calldataSize` actually returns not just the size, but also a boolean telling you whether the type is dynamic or not.  Maybe it should get a different name?  Oh well, that's what I called it for now.  Note, by the way, that all calldata sizes are given in bytes, since calldata works like memory in that it's fundamentally a flat array of bytes, and we never need to reset to a word boundary like in storage.  Of course, all sizes will be multiples of 32, since the ABI is word-based, but they're given in bytes, not the bytes-or-words format of storage lengths.  (A `CalldataPointer` also has the same format as a `MemoryPointer`, aside from the use of `calldata` rather than `memory`.)

Thirdly, recording the calldata in the debugger so that it can be passed to the decoder for decoding.  Now, each call on the EVM callstack, in addition to `address` or `binary`, also contains `data`.  (For creation calls this will be `"0x"`, i.e., empty -- this doesn't matter at present but will in the future.)  The calldata for the initial call of the transaction is fetched from the transaction info during startup; for internal calls (other than creation calls, obviously) the calldata is read from memory when the call is made.  The logic is essentially the same as for reading the binary when a creation call is made; the only difference is what stack position the start and offset arguments are read from.  (This one actually depends on the particular call instruction being used; the 6-argument `STATICCALL` and `DELEGATECALL` have it in a different position from the 7-argument `CALL` and `CALLCODE`.  So a new selector was added to tell us whether the current instruction is a "short call", so that we can always read from the right location.)

That's the essentials.  But this PR also contains a number of other changes, so let's discuss those!

Firstly: In addition to preallocation of structs in storage and calldata, they're now preallocated in *memory* as well.  Allocation in memory, of course, doesn't really consist of anything except filtering out members that are mappings, but, eh, why not I figured.  So now we have allocation in storage, calldata, and in memory.

Let's take a moment to discuss the allocation objects, actually, since these are *almost* the same among the three locations, but not quite.  Memory allocation is the simplest -- for each struct, it gives a pointer for each member. (Except for mappings; these are omitted.  Note that omitting members only occurs in memory allocations, not the others.)  Storage allocation (the already existing case) is similar, except that, firstly, each struct's allocation also contains a field giving the overall size of the struct; and that contracts are allocated as well, not just structs.  A contract's allocation differs in that it lacks the size field, and in that, because contracts can contain constant state variables, the allocation for a member might be a `ConstantDefinitionPointer` rather than a `StoragePointer`.  But this is all old news.  Finally there's calldata allocations.  These also differ from memory allocations in two ways.  Firstly, like in storage, each struct's allocation contains a field for the overall size; but in addition, it also contains a field listing whether the struct is dynamic.

Secondly, in calldata, a struct's allocation might be `null`!  This indicates that this type of struct is not allowed in calldata, because it (somewhere down below) contains a type not allowed in calldata, i.e. a mapping or internal function.  Now, circular structs are also not allowed in calldata, but they will get allocated anyway just because writing code to detect circularity would have been more trouble than it's worth.  (No, they won't throw the allocator into an infinite loop; the rules on what circular structs are allowed at all prevent that.)

Also, I've done some renaming of the decoding functions for consistency.  All decoding functions that involve reading an address, and then acting on that, now all contain `ByAddress` in the name, for clarity.

What else is different?  Well, the debugger now passes `data.current.state` to the debugger, not `data.next.state`.  This does not cause any problems, and indeed fixes a long-standing bug where, when at an assignment statement where a local variable of value type was being assigned to, the new value would be shown before the statement was actually executed.  Now it will display correctly.

That makes this an appropriate time to discuss selector changes, in case we're considering those breaking -- there are several selectors that have been removed in this PR.  Firstly, `data.next` now mostly doesn't exist; only `data.next.state.stack` remains.  (The `createStateSelectors` function is gone, because there was no way to incorporate the new `data.current.state.calldata` into it; `data.next.state.stack` is now just a special case.)  Secondly, I got rid of `data.info.allocations.storage` to instead just have `data.info.allocations` (the result of which contains `storage`, `memory`, and `calldata` members).

Also of note, when adding tests for this PR, I upgraded the Solidity version in the debugger tests from `0.5.1` to `0.5.4`.  Admittedly this may not have been necessary; the tests of actually putting complex data in `calldata` I ended up not including for fear they'd take too long to run, so maybe we don't actually need 0.5.4.  But I did it anyway just in case.

One thing I'll note I'm not really happy about in this PR, is the amount of copypaste I used.  There's a fair bit of repetition (in calldata decoding, and between calldata and memory decoding) that really looks like it should be factored, but I had trouble thinking of a good way to do it that fit in with our existing system, so ultimately I opted to leave the repetition in.  But it still bugs me.

In any case, that's calldata decoding!  Once this is merged, we'll pretty much be able to say that we support decoding for all data types except functions!  (We'll still be missing globally-available builtins, of course, and there will still be storage decoding problems due to failure to keep track, but I'm planning on addressing those soon (to the extent that I can) as well...)

### CODE REVIEW

File: packages/truffle-decoder/lib/allocate/calldata.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  for(const member of definition.members)
+  {
+    let length: number;
+    let dynamicMember: boolean;
+    [length, dynamicMember, allocations] = calldataSizeAndAllocate(member, referenceDeclarations, allocations);
+
+    //vomit on illegal types in calldata
+    if(length === undefined) {
+      allocations[definition.id] = null;
+      return allocations;
```
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/truffle-decoder/lib/allocate/calldata.ts]: Do you think it's worth calling that out in a comment?

File: packages/truffle-debugger/lib/evm/selectors/index.js [hunk 2]
```diff
@@ -125,6 +136,30 @@ function createStepSelectors(step, state = null) {
         }
       ),
 
+      /**
+       * .callData
+       *
+       * data passed to EVM call
+       */
+      callData: createLeaf(
```
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/truffle-debugger/lib/evm/selectors/index.js]: The reason for the different capitalization is that elsewhere we consider `calldata` fairly abstractly, but here it's the data for the call. :)  Like `callAddress` or `callContext`.  So I thought it made sense to go with `callData` here.  But obviously that can be changed for consistency.

File: packages/truffle-debugger/lib/data/selectors/index.js [hunk 3]
```diff
@@ -130,20 +93,22 @@ const data = createSelectorTree({
...
         ref
       ) =>
         forEvmState(definition, ref, {
           referenceDeclarations,
           state,
           mappingKeys,
-          storageAllocations
+          storageAllocations: allocations.storage,
+          memoryAllocations: allocations.memory,
+          calldataAllocations: allocations.calldata
```
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/truffle-debugger/lib/data/selectors/index.js]: Yeah, I think I'll just leave this as-is.

File: packages/truffle-debugger/lib/data/selectors/index.js [hunk 4]
```diff
@@ -261,12 +226,7 @@ const data = createSelectorTree({
     /*
      * data.info.allocations
      */
-    allocations: {
-      /*
-       * data.info.allocations.storage
-       */
-      storage: createLeaf(["/state"], state => state.info.allocations.storage)
-    },
+    allocations: createLeaf(["/state"], state => state.info.allocations),
```
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/truffle-debugger/lib/data/selectors/index.js]: A *dynamic* selector seems like overkill.  But I could still make selectors for each of the individual ones (with the overall one created automatically by `reselect-tree`).

File: packages/truffle-debugger/lib/data/sagas/index.js [hunk 5]
```diff
@@ -39,14 +45,14 @@ function* variablesAndMappingsSaga() {
   let decode = yield select(data.views.decoder);
   let scopes = yield select(data.views.scopes.inlined);
   let referenceDeclarations = yield select(data.views.referenceDeclarations);
-  let allocations = yield select(data.info.allocations.storage);
+  let allocations = (yield select(data.info.allocations)).storage;
```
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/truffle-debugger/lib/data/sagas/index.js]: See comment below

File: packages/truffle-debugger/test/data/more-decoding.js [hunk 6]
```diff
@@ -467,7 +467,7 @@ describe("Further Decoding", function() {
     //get offsets of top-level variables for this contract
     //converting to numbers for convenience
     const startingOffsets = Object.values(
-      Object.values(session.view(data.info.allocations.storage)).filter(
+      Object.values(session.view(data.info.allocations).storage).filter(
```
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/truffle-debugger/test/data/more-decoding.js]: See comment above

File: packages/truffle-debugger/lib/data/sagas/index.js [hunk 7]
```diff
@@ -39,14 +45,14 @@ function* variablesAndMappingsSaga() {
...
   let currentAssignments = yield select(data.proc.assignments);
   let mappedPaths = yield select(data.proc.mappedPaths);
   let currentDepth = yield select(data.current.functionDepth);
   let address = yield select(data.current.address); //may be undefined
   let dummyAddress = yield select(data.current.dummyAddress);
 
-  let stack = yield select(data.next.state.stack);
+  let stack = yield select(data.next.state.stack); //note the use of next!
```
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/truffle-debugger/lib/data/sagas/index.js]: Sure, I could add that I guess.

File: packages/truffle-decoder/lib/allocate/calldata.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/truffle-decoder/lib/allocate/calldata.ts]: Well, I went ahead and did so.

**Comment by haltman-at** [comment]
Latest commit adds back in individual allocation selectors, so `data.info.allocations.storage` is no longer gone, in case anyone outside wanted to make use of that for whatever reason.  But the rest of `data.next` is still removed.

File: packages/truffle-decoder/lib/allocate/calldata.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gnidan) [reply packages/truffle-decoder/lib/allocate/calldata.ts]: Nah I realized that's what you meant by "it vomits". It just surprised me to see.

File: packages/truffle-debugger/lib/evm/selectors/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (gnidan) [reply packages/truffle-debugger/lib/evm/selectors/index.js]: Ah cool yeah that's valid

**Pull request merged**